### ISSUE TITLE
goja panics on TypeScript compiler

### ISSUE THREAD

**Issue by nwidger** [dop251/goja#3]
When I try to run `typescriptServices.js` downloaded from [here](https://raw.githubusercontent.com/Microsoft/TypeScript/master/lib/typescriptServices.js) through `goja`'s `vm.RunString` method it panics.

Here is the small test program I used:

``` go
package main

import (
	"fmt"
	"io/ioutil"
	"log"

	"github.com/dop251/goja"
)

func main() {
	buf, err := ioutil.ReadFile("typescriptServices.js")
	if err != nil {
		log.Fatalln(err)
	}

	vm := goja.New()
	v, err := vm.RunString(string(buf))
	if err != nil {
		log.Fatalln(err)
	}

	fmt.Println(v)

	str := `
interface Person {
    firstName: string;
    lastName: string;
}

function greeter(person: Person) {
    return "Hello, " + person.firstName + " " + person.lastName;
}

var user = { firstName: "Jane", lastName: "User" };

document.body.innerHTML = greeter(user);
`

	v, err = vm.RunString(`ts.compile('` + str + `')`)
	if err != nil {
		log.Fatalln("compile:", err)
	}

	fmt.Println(v)
}
```

Running the test program produces this output:

```
$ go run main.go
panic: Unknown statement type: *ast.DebuggerStatement [recovered]
	panic: Unknown statement type: *ast.DebuggerStatement

goroutine 1 [running]:
panic(0x277880, 0xc42014b940)
	/usr/local/go/src/runtime/panic.go:500 +0x1a1
github.com/dop251/goja.compile.func1(0xc4219c3d40, 0xc4219c3d48)
	/Users/niels/go/src/github.com/dop251/goja/runtime.go:679 +0xc9
panic(0x277880, 0xc42014b940)
	/usr/local/go/src/runtime/panic.go:458 +0x243
github.com/dop251/goja.(*compiler).compileStatement(0xc421db6180, 0x4023a0, 0xc420f3f658, 0xc4219c2700)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:53 +0x186
github.com/dop251/goja.(*compiler).compileStatements(0xc421db6180, 0xc420f3c2c0, 0x4, 0x4, 0xc421f75400)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:562 +0x96
github.com/dop251/goja.(*compiler).compileBlockStatement(0xc421db6180, 0xc420f378f0, 0x27fa00)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:586 +0x54
github.com/dop251/goja.(*compiler).compileStatement(0xc421db6180, 0x402220, 0xc420f378f0, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:16 +0x397
github.com/dop251/goja.(*compiler).compileIfStatement(0xc421db6180, 0xc420f3c200, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:517 +0x543
github.com/dop251/goja.(*compiler).compileStatement(0xc421db6180, 0x4025e0, 0xc420f3c200, 0xc41ff03b00)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:24 +0x3cc
github.com/dop251/goja.(*compiler).compileStatements(0xc421db6180, 0xc420f3fe40, 0x1, 0x1, 0x5f1000)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:567 +0x10c
github.com/dop251/goja.(*compiler).compileBlockStatement(0xc421db6180, 0xc420f37890, 0xffffffffffffff00)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:586 +0x54
github.com/dop251/goja.(*compiler).compileStatement(0xc421db6180, 0x402220, 0xc420f37890, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:16 +0x397
github.com/dop251/goja.(*compiledFunctionLiteral).emitGetter(0xc421f89cc0, 0xa7401)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:785 +0x7bc
github.com/dop251/goja.(*compiler).emitExpr(0xc421db6180, 0x403c60, 0xc421f89cc0, 0xc421f90b01)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:990 +0x8c
github.com/dop251/goja.(*compiler).emitVarSetter.func1(0xc421f90b00)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:352 +0x41
github.com/dop251/goja.(*compiler).emitVarSetter1(0xc421db6180, 0xc42097d91a, 0x6, 0x23911, 0xc4219c2ec0)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:323 +0x408
github.com/dop251/goja.(*compiler).emitVarSetter(0xc421db6180, 0xc42097d91a, 0x6, 0x23911, 0x403c60, 0xc421f89cc0)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:353 +0x7b
github.com/dop251/goja.(*compiledIdentifierExpr).emitSetter(0xc4219c2f70, 0x403c60, 0xc421f89cc0)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:361 +0x5d
github.com/dop251/goja.(*compiler).compileFunction(0xc421db6180, 0xc4200a93d0)
	/Users/niels/go/src/github.com/dop251/goja/compiler.go:416 +0xd5
github.com/dop251/goja.(*compiler).compileFunctions(0xc421db6180, 0xc420f3c340, 0x3, 0x4)
	/Users/niels/go/src/github.com/dop251/goja/compiler.go:312 +0x76
github.com/dop251/goja.(*compiledFunctionLiteral).emitGetter(0xc421f89be0, 0xc4219c3301)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:784 +0x78d
github.com/dop251/goja.(*compiledCallExpr).emitGetter(0xc421f8d580, 0x402300)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:1432 +0x90d
github.com/dop251/goja.(*compiler).compileExpressionStatement(0xc421db6180, 0xc420f40670, 0xc4219c3500)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:594 +0xbe
github.com/dop251/goja.(*compiler).compileStatement(0xc421db6180, 0x4024a0, 0xc420f40670, 0x27fa00)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:18 +0x63d
github.com/dop251/goja.(*compiler).compileStatements(0xc421db6180, 0xc420f3a000, 0x11d, 0x200, 0xc4219c3500)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:562 +0x96
github.com/dop251/goja.(*compiler).compileBlockStatement(0xc421db6180, 0xc420e82cf0, 0x400500)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:586 +0x54
github.com/dop251/goja.(*compiler).compileStatement(0xc421db6180, 0x402220, 0xc420e82cf0, 0x100)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:16 +0x397
github.com/dop251/goja.(*compiledFunctionLiteral).emitGetter(0xc421ece020, 0xc4219c3901)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:785 +0x7bc
github.com/dop251/goja.(*compiledCallExpr).emitGetter(0xc421eca540, 0x402300)
	/Users/niels/go/src/github.com/dop251/goja/compiler_expr.go:1432 +0x90d
github.com/dop251/goja.(*compiler).compileExpressionStatement(0xc421db6180, 0xc420f4a410, 0xc421db3600)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:594 +0xbe
github.com/dop251/goja.(*compiler).compileStatement(0xc421db6180, 0x4024a0, 0xc420f4a410, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/compiler_stmt.go:18 +0x63d
github.com/dop251/goja.(*compiler).compile(0xc421db6180, 0xc421db6140)
	/Users/niels/go/src/github.com/dop251/goja/compiler.go:265 +0x219
github.com/dop251/goja.compile(0x0, 0x0, 0xc420546000, 0x4134b0, 0x0, 0x0, 0x0, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/runtime.go:684 +0x3e2
github.com/dop251/goja.Compile(0x0, 0x0, 0xc420546000, 0x4134b0, 0x0, 0x0, 0x0, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/runtime.go:642 +0x57
github.com/dop251/goja.(*Runtime).RunScript(0xc420096580, 0x0, 0x0, 0xc420546000, 0x4134b0, 0x0, 0x0, 0x0, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/runtime.go:713 +0x52
github.com/dop251/goja.(*Runtime).RunString(0xc420096580, 0xc420546000, 0x4134b0, 0x4136b0, 0xc420546000, 0x4134b0, 0x0)
	/Users/niels/go/src/github.com/dop251/goja/runtime.go:708 +0x51
main.main()
	/Users/niels/go/src/github.com/nwidger/goja-test/main.go:18 +0xfd
exit status 2
```

Here is a copy of the `typescriptServices.js` file I used: [typescriptServices.js.txt](https://github.com/dop251/goja/files/585758/typescriptServices.js.txt).

**Comment by nwidger** [comment]
Looks like it's blowing up on line 2935 character 17 of `typescriptServices.js`:

``` js
debugger;
```

I was able to get the first `vm.RunString` to finish without `panic`ing by adding an empty `case *ast.DebuggerStatement:` to `compiler.compileStatement` in `compiler_stmt.go`, but I doubt this is correct.

**Comment by nwidger** [comment]
I had to modify the test program to properly call the TypeScript compiler methods.  Here's a version that prints out the transpiled TypeScript code (assuming the empty `case` statement has been added to `compiler.compileStatement`:

``` go
package main

import (
	"fmt"
	"io/ioutil"
	"log"

	"github.com/dop251/goja"
)

func main() {
	buf, err := ioutil.ReadFile("typescriptServices.js")
	if err != nil {
		log.Fatalln(err)
	}

	vm := goja.New()
	v, err := vm.RunString(string(buf))
	if err != nil {
		log.Fatalln(err)
	}

	str := `'' +
'interface Person {' +
'    firstName: string;' +
'    lastName: string;' +
'}' +
'' +
'function greeter(person: Person) {' +
'    return "Hello, " + person.firstName + " " + person.lastName;' +
'}' +
'' +
'var user = { firstName: "Jane", lastName: "User" };' +
'' +
'document.body.innerHTML = greeter(user);' +
''`

	v, err = vm.RunString(`
var content = ` + str + `;
var compilerOptions = { module: ts.ModuleKind.System };
var res1 = ts.transpileModule(content, {compilerOptions: compilerOptions, moduleName: 'myModule2'});
var res2 = ts.transpile(content, compilerOptions, /*fileName*/ undefined, /*diagnostics*/ undefined, /*moduleName*/ "myModule1");
res2;
`)
	if err != nil {
		log.Fatalln("compile:", err)
	}

	fmt.Println(v)
}
```

Voila, a working TypeScript to Javascript transpiler in Go!  This is a very cool project, thanks for all the hard work you've done on it so far!

**Comment by clarkmcc** [comment]
Your examples @nwidger were invaluable. 

I ran into issues formatting the string correctly before it get's passed into `ts.transpile` and ended up writing a library for it that we use internally at my company. I just published the library here for anyone that wants a quick an easy way to compile Typescript using goja, https://github.com/clarkmcc/go-typescript:

```go
output, err := typescript.TranspileString("let a: number = 10;", nil) // output: var a = 10;
```

This is a seriously cool project, thank you for all the work you've put into this!
